Lib/typing.py

```diff
@@ -2726,11 +2726,20 @@ def __new__(cls, typename, bases, ns):
             class_getitem = _generic_class_getitem
             nm_tpl.__class_getitem__ = classmethod(class_getitem)
         # update from user namespace without overriding special namedtuple attributes
-        for key in ns:
+        for key, val in ns.items():
             if key in _prohibited:
                 raise AttributeError("Cannot overwrite NamedTuple attribute " + key)
             elif key not in _special and key not in nm_tpl._fields:
-                setattr(nm_tpl, key, ns[key])
+                setattr(nm_tpl, key, val)
+                if hasattr(type(val), "__set_name__"):
+                    try:
+                        type(val).__set_name__(val, nm_tpl, key)
```

Comment:
`type(val).__set_name__` is resolved twice. I propose to resolve it only once:
```py
set_name = getattr(type(val), "__set_name__", _sentinel)
if set_name is not _sentinel:
    set_name(val, nm_tpl, key)
```
or
```py
try:
    set_name = type(val).__set_name__
except AttributeError:
    pass
else:
    set_name(val, nm_tpl, key)
```